file.c

```diff
@@ -236,8 +236,80 @@ int kibosh_read(const char *path UNUSED, char *buf, size_t size, off_t offset,
     } else {
         if (file->type == KIBOSH_FILE_TYPE_NORMAL) {
             fault = kibosh_fs_check_read_fault(fs, file->path);
+
+            // inject read_corrupt fault
+            if (fault >= CORRUPT_ZERO) {
+                int i;
+                double fraction = 0.0;
+                char *suffix = NULL;
+                struct kibosh_fault_base **iter;
+                struct kibosh_fault_read_corrupt *corrupt_fault;
+                for (iter = fs->faults->list; *iter; iter++) {
+                    if (strcmp((*iter)->type, KIBOSH_FAULT_TYPE_READ_CORRUPT) == 0) {
+                        corrupt_fault = (struct kibosh_fault_read_corrupt*)(*iter);
+                        fraction = corrupt_fault->fraction;
+                        suffix = corrupt_fault->suffix;
+                    }
+                }
+
+                int pos = (int) ((1.0 - fraction) * ret);
+
+                switch(fault) {
+                    case CORRUPT_RAND:
+                        for (i=0; i < ret; i++) {
+                            if (random_fraction() <= fraction) {
+                                memset(buf+i, (int) round(random_fraction() * 255.0), 1);
+                            }
+                        }
+                        break;
+
+                    case CORRUPT_ZERO:
+                        for (i=0; i < ret; i++) {
+                            if (random_fraction() <= fraction) {
+                                memset(buf+i, 0, 1);
+                            }
+                        }
+                        break;
+
+                    case CORRUPT_RAND_SEQ:
+                        for (; pos < ret; ++pos) {
+                            memset(buf+pos, (int) round(random_fraction() * 255.0), 1);
```

Comment:
nit: could we do this in a single `memcpy` call instead? Perhaps something like this:

```
int numBytes = ret - pos;
char *randomBuf = malloc(numBytes);
int i;
for (i = 0; i < numBytes; i++)
  randomBuf[i] = round(random_fraction() * 255.0), 1);
memcpy(buf+pos, randomBuf, numBytes);
```

This could be similarly be done for other cases.

Reply:
edited as suggested.